### PR TITLE
[5.5] Sync Str with Stringy 3.0.1

### PR DESCRIPTION
**Pull request by fernandobandeira**
This PR intends to sync our core Str class with the behavior provided by Stringy library.

There's a few things that we should consider:

``` @ ``` was removed from the library so it no longers changes to ``` at ```, the library manually changes ``` @ ``` to ``` - ``` when generating the slug, I kept the previous behavior and moved it to the slug function, this way we can override the array in the future without caring about this and we also don't change the current behavior.

The library now uses locale's to replace characters according to the specific language. This also fixes some issues we had in the past of incorrect behavior, for example:

``` ü ``` previously was being replaced by ``` ue ``` now it is replaced by ``` u ``` and is replaced by ``` ue ``` just when using the ``` de ``` locale.

### CODE REVIEW

File: src/Illuminate/Support/Str.php [hunk 1]
```diff
@@ -33,10 +33,17 @@ class Str
...
      * @param  string  $value
+     * @param  string  $language
      * @return string
      */
-    public static function ascii($value)
+    public static function ascii($value, $language = 'en')
     {
+        $languageSpecific = static::languageSpecificCharsArray($language);
+
+        if ($languageSpecific !== null) {
```
Comment (iwex) [inline src/Illuminate/Support/Str.php]: !is_null() seems more readable

File: src/Illuminate/Support/Str.php [hunk 2]
```diff
@@ -522,59 +532,85 @@ protected static function charsArray()
...
+     * Note: Adapted from Stringy\Stringy.
+     *
+     * @see https://github.com/danielstjules/Stringy/blob/3.0.1/LICENSE.txt
+     *
+     * @param  string  $language
+     * @return array
+     */
+    protected static function languageSpecificCharsArray($language)
+    {
+        static $languageSpecific;
```
Comment (iwex) [inline src/Illuminate/Support/Str.php]: original lib has more actions https://github.com/danielstjules/Stringy/blob/master/src/Stringy.php#L1857

File: src/Illuminate/Support/Str.php (hunk 1 above)
Comment (fernandobandeira) [inline src/Illuminate/Support/Str.php]: Changed 👍

**Comment by iwex** [comment]
@taylorotwell it's WIP, and lacks of additional actions like https://github.com/danielstjules/Stringy/blob/master/src/Stringy.php#L1863

**Comment by fernandobandeira** [comment]
@iwex I've done that https://github.com/laravel/framework/pull/18974/files#diff-d19e1b0b8c742f268aabdbf1595bcfbdR614 just simplified it a bit.

**Comment by fernandobandeira** [comment]
@iwex I've done that I just simplified the function https://github.com/laravel/framework/pull/18974/files#diff-d19e1b0b8c742f268aabdbf1595bcfbdR614

The behavior is the same ``` $charsArray ``` is a local static variable with all languages that are loaded, it contains the languages that were used before and add's new entries to the array whenever you use a new one, i removed this since there's just one language it would be overkill right now...

You might have confused it with the other ``` $charsArray ``` variable that contains all characters...

**Comment by iwex** [comment]
@fernandobandeira thanks for explanation :) Stringy has really messy method

File: src/Illuminate/Support/Str.php [hunk 3]
```diff
@@ -522,59 +532,85 @@ protected static function charsArray()
...
+            ' '    => ["\xC2\xA0", "\xE2\x80\x80", "\xE2\x80\x81", "\xE2\x80\x82", "\xE2\x80\x83", "\xE2\x80\x84", "\xE2\x80\x85", "\xE2\x80\x86", "\xE2\x80\x87", "\xE2\x80\x88", "\xE2\x80\x89", "\xE2\x80\x8A", "\xE2\x80\xAF", "\xE2\x81\x9F", "\xE3\x80\x80", "\xEF\xBE\xA0"],
         ];
     }
+
+    /**
+     * Returns the language specific replacements for the ascii method.
+     *
+     * Note: Adapted from Stringy\Stringy.
+     *
+     * @see https://github.com/danielstjules/Stringy/blob/3.0.1/LICENSE.txt
```
Comment (tillkruss) [inline src/Illuminate/Support/Str.php]: Is this link to the licence intentional, or should this link to the source file?

Comment (fernandobandeira) [inline src/Illuminate/Support/Str.php]: I copied it from the other method that graham created, I guess it's intentional to show that we can freely use this code on our project.

**Pull request merged**